go/enclave/rollupchain/rollup_chain.go

```diff
@@ -386,79 +338,117 @@ func (rc *RollupChain) insertBlockIntoL1Chain(block *types.Block, isLatest bool)
 }
 
 // Updates the L1 and L2 chain heads, and returns the new L2 head hash and the produced batch, if there is one.
-func (rc *RollupChain) updateL1AndL2Heads(block *types.Block) (*common.L2RootHash, *core.Batch, error) {
-	// We extract the rollups from the block.
-	rollupsInBlock := rc.bridge.ExtractRollups(block, rc.storage)
-
-	// We determine whether this is a pre-genesis, genesis, or post-genesis block, and handle the block.
-	blockStage, err := rc.getBlockStage(rollupsInBlock)
+func (rc *RollupChain) updateL1AndL2Heads(block *types.Block, isLatestBlock bool) (*common.L2RootHash, *core.Batch, error) {
+	err := rc.processRollups(block)
 	if err != nil {
-		return nil, nil, fmt.Errorf("could not determine block type. Cause: %w", err)
-	}
-
-	switch blockStage {
-	case PreGenesis:
-		return nil, nil, nil
-	case Genesis:
-		l2Head, err := rc.handleGenesisBlock(block, rollupsInBlock)
-		return l2Head, nil, err
-	case PostGenesis:
-		return rc.handlePostGenesisBlock(block, rollupsInBlock)
+		return nil, nil, fmt.Errorf("could not process rollup in block. Cause: %w", err)
 	}
-	return nil, nil, fmt.Errorf("unrecognised block stage")
-}
 
-// Determines if this is a pre-genesis L2 block, the genesis L2 block, or a post-genesis L2 block.
-func (rc *RollupChain) getBlockStage(rollupsInBlock []*core.Rollup) (BlockStage, error) {
-	_, err := rc.storage.FetchBatchByHeight(0)
+	// We determine whether we have produced a genesis batch yet.
+	isGenesis := false
+	l2Head, err := rc.storage.FetchHeadBatchForBlock(block.ParentHash())
 	if err != nil {
 		if !errors.Is(err, errutil.ErrNotFound) {
-			return -1, fmt.Errorf("could not retrieve genesis rollup. Cause: %w", err)
+			return nil, nil, fmt.Errorf("could not retrieve current head batch. Cause: %w", err)
 		}
+		isGenesis = true
+	}
 
-		// If we haven't stored the genesis rollup and there are no rollups in this block, it cannot be the L2
-		// genesis block.
-		if len(rollupsInBlock) == 0 {
-			return PreGenesis, nil
+	var l2HeadTxReceipts types.Receipts
+	// If this isn't the genesis block, we retrieve the stored receipts.
```

Comment:
This comment still pretty confusing. I don't know how to word it, but think it's more like "if L2 chain already exists" or "if L2 genesis batch exists" to avoid confusion (genesis block sounds like L1 genesis)